### PR TITLE
Fabricator class

### PR DESCRIPTION
**Pull request by MGatner**
**Description**
*This is something I keep needing and finally set out to write a module for it, then realized it might be appropriate for the core framework.*

**Fabricator** is a test utility used to create faked data based on Model definitions. Similar to Laravel's "factories" it creates instances of `$model->returnType` with random data generated by Faker.

**Checklist:**
- [x] Add `create()` method for DB inserts and "faux" inserts
- [x] Write unit tests
- [x] Update User Guide
- [X] Securely signed commits
- [X] Component(s) with PHPdocs
- [X] Conforms to style guide

### CODE REVIEW

**Comment by MGatner** [comment]
@lonnieezell et al, still a fair amount of work to do but please take a look at the concept and weigh in on if you consider this something for the core framework.

**Comment by lonnieezell** [comment]
@MGatner yes, I would love to have something like this. I've wanted it numerous times. :) 

Only catch - it should be able to work for those who don't want to use Composer, unfortunately.  Maybe a super simple class that spits out some semi-random stuff for a handful of common elements if Faker isn't present? 

Another thought is to make it a Project instead. I don't have a problem with those requiring Composer.

**Comment by MGatner** [comment]
I understand the need to support non-Composer installations, but just to push back a little:

1. Faker would be a `require-dev` specification, and we already have a few that aren’t supported by downloads. This would be the first class that requires a Composer package outside of framework tests but it also is very specifically only for testing. I don’t feel the same need to support download-only unit tests as I do for projects, especially since the developer will already need to be downloading content (like PHPUnit). 

2. It wouldn’t be hard to add a non-Composer version that always returns (say) a random string of characters. However the unit tests for this class and the code would suggest to a developer that certain formatters would return valid types (like “email”) which would then fail with our non-Composer version. I could see that being confusing.

**Comment by lonnieezell** [comment]
That's a fair point that testing already requires composer. That's convincing enough for me. :)

**Comment by MGatner** [comment]
I think this is ready! @lonnieezell take a look if you would. And anyone else who sees this, I'd appreciate feedback/suggestions/critiques. A lot harder to change things once they are released so let's make this good up front.

File: system/Test/Fabricator.php [hunk 1]
```diff
@@ -530,7 +531,15 @@ protected function createMock(int $count = null)
 
 		if ($this->model->useSoftDeletes)
 		{
-			$fields[$this->model->deletedField] = $datetime;
+			// Make a 20% chance of returning a deleted item
+			if (rand(1, 5) === 3)
```
Comment (michalsn) [inline system/Test/Fabricator.php:535]: It would be nice to be able to overwrite this somehow. What if I want all my entities to have the same "status"?

Comment (MGatner) [reply system/Test/Fabricator.php:535]: You'd use `setOverrides(['deleted_at' => date('Y-m-d H:i:s')])` or whatever. I'm open to other ideas.

Comment (michalsn) [reply system/Test/Fabricator.php:535]: Fair enough.

**Comment by MGatner** [comment]
Thanks! Re: formatters, I wanted to be sure there was a way people could run distinct tests without modifying models. I will probably always use extended models in Tests\Support with the `fake()` method like I described in the doc, but the option is there. 

Looking over your comments now.

File: system/Test/Fabricator.php [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+	 * @param array|null    $formatters Array of property => formatter
+	 * @param string|null   $locale     Locale for Faker provider
+	 *
+	 * @throws \InvalidArgumentException
+	 */
+	public function __construct($model, array $formatters = null, string $locale = null)
+	{
+		if (is_string($model))
+		{
+			$model = model($model);
```
[the remark this answers is not in the document]
Comment (MGatner) [reply system/Test/Fabricator.php:117]: That’s how I had it originally but then I wondered if that would be a problem with `injectMock` expectations. You wrote `model()` so whatever you think makes the most sense.

File: system/Test/Fabricator.php [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+	 * @var array
+	 */
+	protected $overrides = [];
+
+	/**
+	 * Array of single-use data to override faked versions
+	 *
+	 * @var array|null
+	 */
+	protected $tmpOverrides;
```
[the remark this answers is not in the document]
Comment (MGatner) [reply system/Test/Fabricator.php:93]: I was mimicking Model’s temporary properties which I thought were “tmp___” but I looked hair now and they are as you requested. Will change.

File: system/Test/Fabricator.php (hunk 1 above)
Comment (MGatner) [reply system/Test/Fabricator.php]: Deleted field now has null value by default 4658b613

File: user_guide_src/source/testing/fabricator.rst [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    array(
+        'first'  => "Bobby",
+        'email'  => "melissa.strike@fabricon.us",
+        'phone'  => "525-214-2656 x23546",
+        'avatar' => "http://lorempixel.com/800/400/",
+        'login'  => null,
+    )
+
+``setOverrides()`` can take a second parameter to indicate whether this should be a persistent
+override (default) or only for a single action::
```
[the remark this answers is not in the document]
Comment (MGatner) [reply user_guide_src/source/testing/fabricator.rst]: Clarified expected values and defaults for `setOverrides()` 0b2af265

File: system/Test/Fabricator.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (MGatner) [reply system/Test/Fabricator.php]: Model string values now load a fresh instance. Added a test to demonstrate this c9bc7183

File: system/Test/Fabricator.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (MGatner) [reply system/Test/Fabricator.php]: Property name now has an `e` 6ad992f7

**Comment by MGatner** [comment]
Requested changes implemented!

**Comment by lonnieezell** [comment]
Looks great to me. I'm assuming nothing else you want to do to it? If so - feel free to merge!

**Pull request merged**